File: Frontend/src/services/searchQueryService.ts
// searchQueryService.ts
import api, { PageDTO } from './api';

export interface SearchQuery {
  id: string;
  sentence: string;
  linkCount: number;
  description: string | null;
  created_at: string;
  updatedAt: string;
}

export interface SearchQueryRequest {
  sentence: string;
  linkCount: number;
  description?: string;
}

class SearchQueryService {
  private basePath = '/api/v1/scrape/search_queries';

  /**
   * Get all active search queries
   */
  public async getAllSearchQueries(page: number = 1): Promise<PageDTO<SearchQuery>> {
    return await api.get<PageDTO<SearchQuery>>(`${this.basePath}?page=${page}`);
  }

  /**
   * Get all deleted search queries
   */
  public async getDeletedSearchQueries(page: number = 1): Promise<PageDTO<SearchQuery>> {
    return await api.get<PageDTO<SearchQuery>>(`${this.basePath}/deleted?page=${page}`);
  }

  /**
   * Get all search queries including deleted
   */
  public async getAllSearchQueriesIncludingDeleted(page: number = 1): Promise<PageDTO<SearchQuery>> {
    return await api.get<PageDTO<SearchQuery>>(`${this.basePath}/all?page=${page}`);
  }

  /**
   * Get search query by ID
   */
  public async getSearchQueryById(id: string): Promise<SearchQuery> {
    return await api.get<SearchQuery>(`${this.basePath}/${id}`);
  }

  /**
   * Get deleted search query by ID
   */
  public async getDeletedSearchQueryById(id: string): Promise<SearchQuery> {
    return await api.get<SearchQuery>(`${this.basePath}/deleted/${id}`);
  }

  /**
   * Create a new search query
   */
  public async createSearchQuery(data: SearchQueryRequest): Promise<SearchQuery> {
    return await api.post<SearchQuery>(this.basePath, data as unknown as Record<string, unknown>);
  }

  /**
   * Update a search query
   */
  public async updateSearchQuery(id: string, data: SearchQueryRequest): Promise<SearchQuery> {
    return await api.put<SearchQuery>(`${this.basePath}/${id}`, data as unknown as Record<string, unknown>);
  }

  /**
   * Restore a soft-deleted search query
   */
  public async restoreSearchQuery(id: string): Promise<void> {
    await api.put<void>(`${this.basePath}/${id}/restore`, {});
  }

  /**
   * Soft delete a search query
   */
  public async deleteSearchQuery(id: string): Promise<void> {
    await api.delete(`${this.basePath}/${id}`);
  }

  /**
   * Permanently delete a search query
   */
  public async fullDeleteSearchQuery(id: string): Promise<void> {
    await api.delete(`${this.basePath}/${id}/full_delete`);
  }
}

const searchQueryService = new SearchQueryService();
export default searchQueryService;

File: Backend/src/main/java/ir/netpick/platform/mailmine/service/base/SearchQueryService.java
package ir.netpick.platform.mailmine.service.base;

import java.util.Objects;
import java.util.UUID;
import ir.netpick.platform.core.PageDTO;
import ir.netpick.platform.core.constants.GeneralConstants;
import ir.netpick.platform.mailmine.dto.SearchQueryResponse;
import ir.netpick.platform.mailmine.mapper.SearchQueryDTOMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.validation.annotation.Validated;

import ir.netpick.platform.core.exception.ResourceNotFoundException;
import ir.netpick.platform.core.utils.PageDTOMapper;
import ir.netpick.platform.mailmine.dto.SearchQueryRequest;
import ir.netpick.platform.mailmine.model.SearchQuery;
import ir.netpick.platform.mailmine.repository.SearchQueryRepository;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.RequiredArgsConstructor;

@Slf4j
@Validated
@Service
@RequiredArgsConstructor
public class SearchQueryService {

    private final SearchQueryRepository searchQueryRepository;
    private final SearchQueryDTOMapper searchQueryDTOMapper;

    public boolean isEmpty() {
        return searchQueryRepository.count() == 0;
    }

    @Cacheable(value = "searchQueries", key = "'page-' + #pageNumber")
    public PageDTO<SearchQueryResponse> allSearchQueries(int pageNumber) {
        Pageable pageable = PageRequest.of(pageNumber - 1, GeneralConstants.PAGE_SIZE,
                Sort.by("createdAt").descending());
        Page<SearchQuery> page = searchQueryRepository.findByDeletedFalse(pageable);
        return PageDTOMapper.map(page, searchQueryDTOMapper);
    }

    public PageDTO<SearchQueryResponse> deletedSearchQueries(int pageNumber) {
        Pageable pageable = PageRequest.of(pageNumber - 1, GeneralConstants.PAGE_SIZE,
                Sort.by("createdAt").descending());
        Page<SearchQuery> page = searchQueryRepository.findByDeletedTrue(pageable);
        return PageDTOMapper.map(page, searchQueryDTOMapper);

    }

    public PageDTO<SearchQueryResponse> allSearchQueriesIncludingDeleted(int pageNumber) {
        Pageable pageable = PageRequest.of(pageNumber - 1, GeneralConstants.PAGE_SIZE,
                Sort.by("createdAt").descending());
        Page<SearchQuery> page = searchQueryRepository.findAll(pageable);
        return PageDTOMapper.map(page, searchQueryDTOMapper);
    }

    @Cacheable(value = "searchQuery", key = "#id")
    public SearchQuery getSearchQuery(@NotNull UUID id) {
        return searchQueryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("SearchQuery with ID [%s] not found.".formatted(id)));
    }

    public SearchQuery getSearchQueryIncludingDeleted(@NotNull UUID id) {
        return searchQueryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("SearchQuery with ID [%s] not found.".formatted(id)));
    }

    @CacheEvict(value = { "searchQuery", "searchQueries" }, allEntries = true)
    public SearchQuery createSearchQuery(@Valid @NotNull SearchQueryRequest request) {
        SearchQuery searchQuery = new SearchQuery(
                request.sentence(),
                request.description());
        SearchQuery saved = searchQueryRepository.save(searchQuery);
        log.info("Created SearchQuery with ID: {}", saved.getId());
        return saved;
    }

    @CacheEvict(value = { "searchQuery", "searchQueries" }, allEntries = true)
    public SearchQuery updateSearchQuery(@NotNull UUID id, @Valid @NotNull SearchQueryRequest request) {
        SearchQuery existing = getSearchQuery(id);

        boolean changed = false;

        if (Objects.nonNull(request.sentence()) && !Objects.equals(request.sentence(), existing.getSentence())) {
            existing.setSentence(request.sentence());
            changed = true;
        }

        if (Objects.nonNull(request.description())
                && !Objects.equals(request.description(), existing.getDescription())) {
            existing.setDescription(request.description());
            changed = true;
        }

        if (!changed) {
            log.warn("No changes found in SearchQuery update request for ID: {}", id);
            return existing; // Or throw if strict
        }

        SearchQuery saved = searchQueryRepository.save(existing);
        log.info("Updated SearchQuery with ID: {}", id);
        return saved;
    }

    @CacheEvict(value = { "searchQuery", "searchQueries" }, allEntries = true)
    public void softDeleteSearchQuery(@NotNull UUID id) {
        if (!searchQueryRepository.existsById(id)) {
            throw new ResourceNotFoundException("SearchQuery with ID [%s] not found.".formatted(id));
        }
        searchQueryRepository.softDelete(id);
        log.info("Soft deleted SearchQuery with ID: {}", id);
    }

    @CacheEvict(value = { "searchQuery", "searchQueries" }, allEntries = true)
    public void restoreSearchQuery(@NotNull UUID id) {
        if (!searchQueryRepository.existsById(id)) {
            throw new ResourceNotFoundException("SearchQuery with ID [%s] not found.".formatted(id));
        }
        searchQueryRepository.restore(id);
        log.info("Restored SearchQuery with ID: {}", id);
    }

    @CacheEvict(value = { "searchQuery", "searchQueries" }, allEntries = true)
    public void deleteSearchQuery(@NotNull UUID id) {
        if (!searchQueryRepository.existsById(id)) {
            throw new ResourceNotFoundException("SearchQuery with ID [%s] not found.".formatted(id));
        }
        searchQueryRepository.deleteById(id);
        log.info("Deleted SearchQuery with ID: {}", id);
    }
}








